bip-0352.mediawiki

```diff
@@ -0,0 +1,472 @@
+
+''' P2TR '''
+
+The sender MUST use the private key corresponding to the taproot output key (i.e. the tweaked private key for a key path spend). This can be a single private key or an aggregate key (e.g. taproot outputs using MuSig or FROST)<ref name="musig_frost_support">'''Are key aggregation techniques like FROST and MuSig supported?''' While we do not recommend it due to lack of a security proof (except if all participants are trusted or are the same entity), any taproot output able to do a key path theoretically is supported. Any offline key aggregation technique can be used, such as FROST or MuSig. This would require participants to perform the ECDH step collaboratively e.g. ''ECDH = a<sub>0</sub>·B<sub>scan</sub> + a<sub>1</sub>·B<sub>scan</sub> + ... + a<sub>t</sub>·B<sub>scan</sub>'' and ''P = B<sub>spend</sub> + sha256(outpoints_hash·ECDH || 0)·G''. Additionally, it may be necessary for the participants to provide a DLEQ proof to ensure they are not acting maliciously.</ref>. If this key is not available, the output cannot be included as an input to the transaction. The receiver always uses the taproot output key when scanning, regardless of whether the taproot output is using a key path spend or a script path spend<ref name="why_always_output_pubkey">''' Why not skip all taproot script path spends? ''' This causes malleability issues for CoinJoins. If the silent payments protocol skipped taproot script path spends, this would allow an attacker to join a CoinJoin round, participate in deriving the silent payment address using the tweaked private key for a key path spend, and then broadcast their own version of the transaction using the script path spend. If the receiver were to only consider key path spends, they would skip the attacker's script path spend input when deriving the shared secret and not be able to find the funds. Additionally, there may be scenarios where a sender has access to the key path private key but spends the output using the script path.</ref>.
+
+The one exception is script path spends that use NUMS point ''H'' as their internal key (where ''H = lift_x(0x50929b74c1a04954b78b4b6035e97a5e078a5a0f28ec96d547bfee9ace803ac0)'' which is constructed by taking the hash of the standard uncompressed encoding of the secp256k1 base point ''G'' as X coordinate, see [https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#constructing-and-spending-taproot-outputs BIP341: Constructing and spending Taproot outputs] for more details), in which case the output will be skipped for the purposes of shared secret derivation<ref name="why_ignore_h">'''Why skip outputs with H as the internal taproot key?''' If use cases get popularized where the taproot key path cannot be used, these outputs can still be included without getting in the way of making a silent payment, provided they specifically use H as their internal taproot key.</ref>.
```

Comment:
If I understand correctly, this requires all participants in a CoinJoin to reveal their internal key and taproot tweak, so that others can check the internal key is not _H_. If that's correct, can this be explicitly stated?

Reply:
Not quite. In order to generate a silent payment output you need the cooperation of all input owners. They'd have to perform Diffie-Hellman and send you the result (together with a DLEQ), but if their internal key is not known then they wouldn't be able to, so the fact that they succeeded automatically tells you that their internal key is not H (or some other NUMS). If the internal key is in fact H and they want to include the input, then they do have to reveal it and then the input can be safely ignored.

Reply:
You're right, I didn't realize that using any NUMS as the internal key means the sender can't do ECDH.

It would be better for privacy to allow any _H + rG_ as the internal key when the input should be skipped, but I'm not sure it's possible for the receiver to detect this tweak without too big of a performance hit.

Reply:
Such a tweak would indeed be impossible to detect as that's the point of doing it in the first place - hiding the fact that you couldn't sign with the internal key from the outside world.

Thanks for asking such in-depth questions, it really helps with review.

Reply:
The sender would of course have to communicate _r_ to the receiver in some way so that they can verify the tweak. I initially had a scheme in mind that was inefficient, but I later realized it doesn't actually work.

Reply:
If the sender and receiver can communicate then you no longer need silent payments and can follow a much simpler protocol. Either the recipient simply communicates a fresh key to the sender, or the sender tweaks the recipient key with some secret value and communicates this value to the recipient.

Reply:
By communicate I meant through the transaction itself, same as the rest of the Silent Payments protocol. As a trivial example, you could have your script begin with `OP_FALSE OP_IF <r> OP_ENDIF`, but that reveals _r_ to everyone, not just the recipient. I'm not sure it's even possible to do it privately.

Reply:
Right, yeah that's possible in a private way but it requires Diffie-Hellman to blind the value and you end up taking up additional block space and reducing the anonymity set because the tx will stand out. We considered something along those lines as a fallback for when none of the inputs were silent payment compatible, but decided against it for those reasons.